Thank you for working on this! Had a quick test, and hit a couple of issues:

## Duplicate `writeMinecraftClasspathXyz`  task
I have a split source-set mod, with both a `main` (for common code) and `client` (for client code). However, because [runs are re-configured for each source-set](https://github.com/neoforged/NeoGradle/blob/c35e04546f7f3d0fb36e0cf58367d47dacdca29f/common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java#L139-L146), this causes the `writeMinecraftClasspath` task to be created multiple times.

<details><summary>Error message + stack trace</summary>

```
org.gradle.api.internal.tasks.DefaultTaskContainer$DuplicateTaskException: Cannot add task 'writeMinecraftClasspathClient' as a task with that name already exists.
        at org.gradle.api.internal.tasks.DefaultTaskContainer.failOnDuplicateTask(DefaultTaskContainer.java:257)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.registerTask(DefaultTaskContainer.java:404)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.register(DefaultTaskContainer.java:381)
        at net.neoforged.gradle.userdev.runtime.definition.UserDevRuntimeDefinition.buildRunInterpolationData(UserDevRuntimeDefinition.java:135)
        at net.neoforged.gradle.common.runtime.definition.CommonRuntimeDefinition.configureRun(CommonRuntimeDefinition.java:174)
        at net.neoforged.gradle.common.CommonProjectPlugin.lambda$null$8(CommonProjectPlugin.java:142)
        at net.neoforged.gradle.common.CommonProjectPlugin.lambda$null$9(CommonProjectPlugin.java:142)
        at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
        at net.neoforged.gradle.common.CommonProjectPlugin.lambda$null$10(CommonProjectPlugin.java:139)
        at net.neoforged.gradle.common.CommonProjectPlugin.lambda$applyAfterEvaluate$11(CommonProjectPlugin.java:131)
        at org.gradle.internal.extensibility.ExtensionsStorage$ExtensionHolder.configure(ExtensionsStorage.java:173)
        at org.gradle.internal.extensibility.ExtensionsStorage.configureExtension(ExtensionsStorage.java:64)
        at org.gradle.internal.extensibility.DefaultConvention.configure(DefaultConvention.java:195)
        at net.neoforged.gradle.common.CommonProjectPlugin.applyAfterEvaluate(CommonProjectPlugin.java:131)
```

</details>

## `ObjectInstantiationException` for `ConfigurationRunDependencyImpl`:
Trying to use the example code in the OP (`dependencies { runtime(configuration(project.configurations["minecraftLibrary"])) }`), results in the following error:

<details><summary>Error message + stack trace</summary>

```
Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type net.neoforged.gradle.common.runs.run.ConfigurationRunDependencyImpl.
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
        at org.gradle.api.internal.model.DefaultObjectFactory.newInstance(DefaultObjectFactory.java:89)
        at net.neoforged.gradle.common.runs.run.DependencyHandlerImpl.configuration(DependencyHandlerImpl.java:77)
        at Build_gradle$5$1$2.execute(build.gradle.kts:54)
        at Build_gradle$5$1$2.execute(build.gradle.kts:53)
        at net.neoforged.gradle.dsl.common.runs.run.Run.dependencies(Run.groovy)
        at net.neoforged.gradle.dsl.common.runs.run.Run.dependencies(Run.groovy)
        at Build_gradle$5$1.execute(build.gradle.kts:53)
        at Build_gradle$5$1.execute(build.gradle.kts:39)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:123)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:110)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:107)
        at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:45)
        at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:45)
        at org.gradle.internal.ImmutableActionSet$SetWithFewActions.execute(ImmutableActionSet.java:285)
        at org.gradle.api.internal.DefaultDomainObjectCollection.doAdd(DefaultDomainObjectCollection.java:262)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection.doAdd(DefaultNamedDomainObjectCollection.java:113)
        at org.gradle.api.internal.DefaultDomainObjectCollection.add(DefaultDomainObjectCollection.java:256)
        at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.tryCreate(DefaultNamedDomainObjectCollection.java:944)
        ... 261 more
Caused by: org.gradle.internal.instantiation.ClassGenerationException: Could not generate a decorated class for type ConfigurationRunDependencyImpl.
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generateUnderLock(AbstractClassGenerator.java:247)
        at org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$AbstractCrossBuildInMemoryCache.get(DefaultCrossBuildInMemoryCacheFactory.java:130)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generate(AbstractClassGenerator.java:173)
        at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator.generate(AsmBackedClassGenerator.java:118)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.lambda$forType$0(Jsr330ConstructorSelector.java:56)
        at org.gradle.cache.Cache.lambda$get$0(Cache.java:31)
        at org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$AbstractCrossBuildInMemoryCache.get(DefaultCrossBuildInMemoryCacheFactory.java:130)
        at org.gradle.cache.Cache.get(Cache.java:31)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forType(Jsr330ConstructorSelector.java:53)
        at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forParams(Jsr330ConstructorSelector.java:48)
        at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:61)
        ... 288 more
Caused by: java.lang.IllegalArgumentException: Cannot have abstract method ProjectAssociatedDSLElement.getProject().
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.assertNotAbstract(AbstractClassGenerator.java:388)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.inspectType(AbstractClassGenerator.java:314)
        at org.gradle.internal.instantiation.generator.AbstractClassGenerator.generateUnderLock(AbstractClassGenerator.java:216)
        ... 298 more
```

</details>

--- a/common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java
+++ b/common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java
@@ -35,7 +35,9 @@ import org.gradle.plugins.ide.eclipse.EclipsePlugin;
 import org.gradle.plugins.ide.idea.IdeaPlugin;
 import org.jetbrains.gradle.ext.IdeaExtPlugin;
 
+import java.util.HashSet;
 import java.util.Optional;
+import java.util.Set;
 
 public class CommonProjectPlugin implements Plugin<Project> {
     
@@ -136,13 +138,20 @@ public class CommonProjectPlugin implements Plugin<Project> {
                 
                 if (run.getConfigureFromDependencies().get()) {
                     final RunImpl runImpl = (RunImpl) run;
+                    
+                    final Set<CommonRuntimeDefinition<?>> definitionSet = new HashSet<>();
+                    
                     runImpl.getModSources().get().forEach(sourceSet -> {
                         try {
                             final Optional<CommonRuntimeDefinition<?>> definition = TaskDependencyUtils.findRuntimeDefinition(project, sourceSet);
-                            definition.ifPresent(def -> def.configureRun(runImpl));
+                            definition.ifPresent(definitionSet::add);
                         } catch (MultipleDefinitionsFoundException e) {
                             throw new RuntimeException("Failed to configure run: " + run.getName() + " there are multiple runtime definitions found for the source set: " + sourceSet.getName(), e);
                         }
+                    });
+                    
+                    definitionSet.forEach(definition -> {
+                       definition.configureRun(runImpl);
                     });
                 }
             }

--- a/common/src/main/java/net/neoforged/gradle/common/runs/run/ConfigurationRunDependencyImpl.java
+++ b/common/src/main/java/net/neoforged/gradle/common/runs/run/ConfigurationRunDependencyImpl.java
@@ -13,6 +13,8 @@ import javax.inject.Inject;
 
 public abstract class ConfigurationRunDependencyImpl implements RunDependency {
 
+    private final Project project;
+    
     @Inject
     public ConfigurationRunDependencyImpl(Project project, Configuration dependency) {
         getIdentity().convention(dependency.toString());
@@ -21,8 +23,14 @@ public abstract class ConfigurationRunDependencyImpl implements RunDependency {
             final ConfigurableFileCollection files = project.files();
             return files.from(resolvedConfiguration.getFiles());
         }));
+        this.project = project;
     }
-
+    
+    @Override
+    public Project getProject() {
+        return project;
+    }
+    
     @Override
     public abstract ConfigurableFileCollection getDependency();
 

--- a/common/src/main/java/net/neoforged/gradle/common/runs/run/RunDependencyImpl.java
+++ b/common/src/main/java/net/neoforged/gradle/common/runs/run/RunDependencyImpl.java
@@ -13,6 +13,8 @@ import javax.inject.Inject;
 
 public abstract class RunDependencyImpl implements RunDependency {
 
+    private final Project project;
+    
     @Inject
     public RunDependencyImpl(Project project, Dependency dependency) {
         getIdentity().convention(dependency.toString());
@@ -22,8 +24,14 @@ public abstract class RunDependencyImpl implements RunDependency {
             final ConfigurableFileCollection files = project.files();
             return files.from(resolvedConfiguration.getFiles());
         }));
+        this.project = project;
     }
-
+    
+    @Override
+    public Project getProject() {
+        return project;
+    }
+    
     @Override
     public abstract ConfigurableFileCollection getDependency();
 

